### PR TITLE
Fix metric attribute lookup

### LINKED ISSUE #8179: `.log` no longer supports nested metric attributes (e.g. if stored in `ModuleDict`)
## 🐛 Bug

`Metric` objects stored in a `ModuleDict` used to be supported but now gives the following error:
```
pytorch_lightning.utilities.exceptions.MisconfigurationException: Could not find the `LightningModule` attribute for the `torchmetrics.Metric` logged. You can fix this by setting an attribute for the metric in your `LightningModule`.
```

More info in PyTorchLightning/lightning-flash#481

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

Fixes #8179 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
LGTM ! Great catch !

**Review by ananyahjha93 — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

File: tests/metrics/test_metric_lightning.py [hunk 1]
```diff
@@ -192,3 +210,78 @@ def training_epoch_end(self, outputs):
...
+            self.a_metric = SumMetric()
+
+        def training_step(self, *args):
+            metric = SumMetric()
+            self.log("foo", metric)
+
+    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=1)
+    model = TestModel()
+    with pytest.raises(MisconfigurationException, match=escape("where `name` is one of ['a_metric']")):
+        trainer.fit(model)
```
Comment (ananthsub) [inline tests/metrics/test_metric_lightning.py:245]: @carmocca why do we need this dependency on the attribute name in the module vs the key name used for publishing? if it's solely for restoration, could we think of other approaches? this style of metrics logging is a very common pattern for us

Comment (mooey5775) [reply tests/metrics/test_metric_lightning.py:245]: Not @carmocca, but I'll take a shot at this - I don't think there is a dependency on key name for publishing. This test is failing because the metric being logged isn't the same instance as the `self.a_metric` metric that's an attribute in the module.

`test_log_metric_dict` below logs some metrics that have key name != attribute name and still passes

Comment (carmocca) [reply tests/metrics/test_metric_lightning.py:245]: @mooey5775 is correct.

This is to match the logged metric with the metric attribute in the `LightningModule` so the state can be restored.

> could we think of other approaches?

Open to ideas!

**Comment by kandluis** [comment]
Let's say I have some metrics which I don't want to restore/track (and therefore, this metric attribute is no needed).

What's the recommended way to log these metrics in that case?

**Comment by carmocca** [comment]
> which I don't want to restore/track (and therefore, this metric attribute is no needed).

But it's still a stateful metric? Can't change to a functional counterpart?

> What's the recommended way to log these metrics in that case?

There isn't currently for this as the `log` method requires it if a `torchmetrics.Metric` is passed.
Can you open a feature request about it?
